lib/matplotlib/patches.py

```diff
@@ -989,6 +989,78 @@ def set_path(self, path):
         self._path = path
 
 
+class StepPatch(PathPatch):
+    """An unclosed step path patch."""
+
+    @docstring.dedent_interpd
+    def __init__(self, values, edges, *,
+                 orientation='vertical', baseline=0, **kwargs):
+        """
+        Parameters
+        ----------
+        values : array-like
+            An array of y-values.
+
+        edges : array-like
+            A array of x-value edges, with ``len(edges) == len(vals) + 1``,
+            between which the curve takes on vals values.
+
+        orientation : {'vertical', 'horizontal'}, default: 'vertical'
+
+        baseline : float or None, default: 0
+            Determines starting value of the bounding edges or when
+            ``fill=True``, position of lower edge.
+
+        Other valid keyword arguments are:
+
+        %(Patch)s
+        """
+        self.baseline = baseline
+        self.orientation = orientation
+        self._edges = np.asarray(edges)
+        self._values = np.asarray(values)
+        verts, codes = self._update_data()
+        path = Path(verts, codes)
+        super().__init__(path, **kwargs)
+
+    def _update_data(self):
+        if np.isnan(np.sum(self._edges)):
+            raise ValueError('Nan values in "edges" are disallowed')
+        if self._edges.size - 1 != self._values.size:
+            raise ValueError('Size mismatch between "values" and "edges". '
+                             "Expected `len(values) + 1 == len(edges)`, but "
+                             "they are or lengths {} and {}".format(
+                                 self._edges.size, self._values.size)
+                             )
+        verts, codes = [], []
+        for idx0, idx1 in cbook.contiguous_regions(~np.isnan(self._values)):
+            x = np.repeat(self._edges[idx0:idx1+1], 2)
+            y = np.repeat(self._values[idx0:idx1], 2)
+            if self.baseline is not None:
+                y = np.hstack((self.baseline, y, self.baseline))
+            else:
+                y = np.hstack((y[0], y, y[-1]))
+            if self.orientation == 'vertical':
+                xy = np.column_stack([x, y])
+            else:
+                xy = np.column_stack([y, x])
+            verts.append(xy)
+            codes.append(np.array([Path.MOVETO] + [Path.LINETO]*(len(xy)-1)))
+        return np.vstack(verts), np.hstack(codes)
+
+    def set_edges(self, edges):
```

Comment:
Do you need to catch that the edges and values are different sized?  I guess you do it in `_update_data`.  Is there any point in a `set_data` method that allows both to be set and changed at once.  Otherwise the user cannot change the number of bins.

Reply:
huge :+1: from me to be able to change the number of bins via a `set_data` method

Reply:
This is a good catch. In keeping with the previous comment about no duplicate methods, should we simply have `set_data` instead of both vals/edges with signature:
```
s.set_data([1,2,3])
s.set_data(None, [0,1,2,3])
s.set_data([1,2,3], [0,1,2,3])
```
@timhoffm since this is one more API question

Reply:
I don't think having all three is a problem, though you could remove any duplicated code and just call `set_data` with `set_edges` and `set_values`.